distml/operator/jax_operator.py

```diff
@@ -0,0 +1,338 @@
+        targets_class = jnp.argmax(targets, axis=1)
+
+        acc = jnp.mean(prediction_class == targets_class)
+        samples_num = targets.shape[0]
+
+        return {
+            "val_loss": loss.item(),
+            "val_accuracy": acc.item(),
+            "samples_num": samples_num
+        }
+
+    def get_parameters(self, cpu):
+        """get the flatten parameters."""
+        params = self.get_params(self.opt_state)
+        flatten_params, tree = tree_flatten(params)
+        if not hasattr(self, "tree"):
+            self.tree = tree
+
+        if cpu:
+            flatten_params = list(map(np.asarray, flatten_params))
+        return flatten_params
+
+    def get_named_parameters(self, cpu):
+        """Get the named parameters.
+
+        In jax, we need to construct a dict to contain the parameters.
+        """
+        params = self.get_parameters(cpu)
+        if hasattr(self, "preset_keys"):
+            dict_params = {
+                name: p
+                for name, p in zip(self.preset_keys, params)
+            }
+        else:
+            dict_params = {f"{idx}": p for idx, p in enumerate(params)}
+        return dict_params
+
+    # TODO(HUI): used in load states or load parameters
+    def set_parameters(self, new_params):
+        """Use new parameters to replace model parameters.
+
+        In jax, we need to construct a dict to contain the parameters.
+
+        Args:
+            new_params (dict): New parameters to updates the current model.
+        """
+        assert isinstance(new_params, dict)
+
+        keys, new_params = unzip2(
+            sorted(new_params.items(), key=lambda d: int(d[0])))
+        self.preset_keys = keys
+
+        if not hasattr(self, "tree"):
+            self.tree = tree_structure(self.get_params(self.opt_state))
+
+        states_flat, tree, subtrees = self.opt_state
+
+        states = map(tree_unflatten, subtrees, states_flat)
+
+        def update(param, state):
+            new_state = param, *state[1:]
+            return new_state
+
+        new_states = map(update, new_params, states)
+
+        new_states_flat, new_subtrees = unzip2(map(tree_flatten, new_states))
+
+        if not new_subtrees:
+            raise RuntimeError("subtrees of new params is empty.")
+        for idx, (subtree, new_subtree) in enumerate(
+                zip(subtrees, new_subtrees)):
+            if new_subtree != subtree:
+                msg = (
+                    "input structur did not match the save params struture. "
+                    "input {} and output {}.")
+                raise TypeError(msg.format(subtree, new_subtree))
+
+        self.opt_state = OptimizerState(new_states_flat, tree, new_subtrees)
+
+    def reset_optimizer_for_params(self, params):
+        keys, params = unzip2(sorted(params.items(), key=lambda d: int(d[0])))
+        self.tree = tree_structure(params)
+        self.opt_state = self.opt_init(params)
+
+    def ones(self, shape, cpu=True):
```

Comment:
All methods below seem to be class-agonostic (staticmethods).
Maybe we should lift the out and put them in a file called `jax_util.py`?

Reply:
These methods will be called in replicas/actors.  If put these in `jax_util.py`, the replicas/actors need to realize the operator belongs to which ml framework. That's not what we want.  But we can use `@staticmethod` to decorate these methods.

Reply:
I don't know how to deal with these functions now. I'm going to remove them as they are not used by the AllReduceStrategy. Let's revisit them in the next PS pull request.